### ISSUE TITLE
dvc: expand on `--no-pager` option

### ISSUE THREAD

**Issue by jorgeorpinel** [iterative/dvc#5793]
1. ~~Make it the default for `dvc exp show`?~~ No longer needed after #5906
2. Include it in `dvc dag`?
3. Maybe put it directly in `dvc --no-pager` so it works for any commands that need it in the future.
4. Any other existing commands that need it? E.g. `list` or `stage list`

Rel. #4060

**Comment by pmrowla** [comment]
What should the behavior be for dags that are too wide to display in the user's terminal? Wrapping lines (or truncating it) eliminates the usefulness of the output entirely here, it's not the same as a table where we can merge/hide columns until the table is small enough to fit.

**Comment by dmpetrov** [comment]
Should we do `--no-pager` as a default in both of the cases? I use `--no-pager` more often than the default one.

**Comment by jorgeorpinel** [comment]
> What should the behavior be for dags that are too wide

Good Q. Maybe just truncate it like I think you are saying we do for tables?

> Should we do --no-pager as a default in both of the cases?

I also use `--no-pager` all the time haha. Or at least give it a short flag. BTW `git` has `--no-pager` in the top command i.e. `git --no-pager diff`; another option to consider.

**Comment by shcheklein** [comment]
> More options to consider.

detect if you have enough space or not and apply pager if not? that's what git does on my machine, but as we discussed with Ruslan it's not clear how/why it is happening.

**Comment by dmpetrov** [comment]
For most of my scenarios, it is never enough space for the whole table. But I still prefer `--no-pager`.

**Comment by jorgeorpinel** [comment]
BTW I renamed this issue and rewrote the description to reflect what we're actually talking about. To summarize, I propose:

Step 1. Make it the default for `exp show`
Step 2. Put it directly in `dvc --no-pager` and enable if for `dag` too (but not as default).

Qs:
- How to handle long lines? I say truncate them for now.
- Any other commands that need it? E.g. list or stage list

**Comment by dberenbaum** [comment]
> detect if you have enough space or not and apply pager if not? that's what git does on my machine, but as we discussed with Ruslan it's not clear how/why it is happening.

Seems like Git is using `less -FRX` by default to do this: https://git-scm.com/docs/git-config#Documentation/git-config.txt-corepager. It would be great to do the same in dvc and make this the default.

>     * How to handle long lines? I say truncate them for now.

👍 

>     * Any other commands that need it? E.g. `list`

Ideally, we can abstract and set all of this at the top level as part of the UI overhaul that's ongoing. Thought @skshetry?

**Comment by skshetry** [comment]
> > detect if you have enough space or not and apply pager if not? that's what git does on my machine, but as we discussed with Ruslan it's not clear how/why it is happening.
> 
> Seems like Git is using `less -FRX` by default to do this: https://git-scm.com/docs/git-config#Documentation/git-config.txt-corepager. It would be great to do the same in dvc and make this the default.

good to finally know the cause 😄, anyway it adds support to my proposal https://github.com/iterative/dvc/issues/4060#issuecomment-769846201 ;-)

**Comment by skshetry** [comment]
I'm working on keeping the behavior same as `git` as proposed on https://github.com/iterative/dvc/issues/4060#issuecomment-769846201. But very few people will be able to see this as most users will have `PAGER` or `LESS` set.

**Comment by dmpetrov** [comment]
> I'm working on keeping the behavior same as `git` as proposed on [#4060 (comment)](https://github.com/iterative/dvc/issues/4060#issuecomment-769846201).

It would be great to have similar to Git behavior. However, it is not limiting by --no-pager/less mode. In Git when I quite from the less mode (like `git log`) I still see the output. I would not need `--no-pager` option of `exp run` if this would be the case. But the current default mode clears out the output - that's the problem.

@skshetry could you please clarify what users will see after quitting the command in less mode?

**Comment by skshetry** [comment]
@dmpetrov, well I have two ideas:

1. Print the table in a pager mode and in a non-pager mode after the pager exits. This way, you will see a non-collapsed form in the table and collapsed form after it exits.
2. `DVC_PAGER="less -FRSX~" dvc exp show` -> This way, if the content can fit within the screen (vertically and horizontally), no pager will be used.  Otherwise, the pager will be shown where the user's prompt was (without clearing the screen to the top). After that, the content will still be on the user's screen. \

2) has a disadvantage that when the user scrolls, it will leave the empty patch in the user's terminal. So, if we need to show both, I'd prefer the 1), though it feels like a hack.

thoughts @pmrowla?

**Comment by pmrowla** [comment]
> It would be great to have similar to Git behavior. However, it is not limiting by --no-pager/less mode. In Git when I quite from the less mode (like `git log`) I still see the output.

@dmpetrov 

This is not actually default Git behavior (when I quit from `git log` inside a pager I do not see the output). It sounds to me like something in your machine/environment is configuring the system pager to be `less -X`.

Can you run
```
echo $PAGER
```
in a terminal and see what it outputs?

---

IMO we should leave the existing defaults, and if the user wants the `-X` behavior, they can configure `DVC_PAGER` as needed, the same way it works with configuring `PAGER` for git.

If anything, we could maybe consider just respecting the system `PAGER` instead of using our own env var.

**Comment by pmrowla** [comment]
You can disregard my last comment, I see where Dave linked the git docs w/the info about their defaults now.

Using the same `less -FRSX` defaults makes sense to me

@skshetry

**Comment by skshetry** [comment]
@pmrowla, `dvc exp show` needs `-R` to work properly. `PAGER` is usually set to `less` and `LESS` is already set up in my machine with `-R` (by ohmyzsh?), but we cannot say the same for everyone. So it feels like we cannot just honour `$LESS`.

I'd propose we detect if the pager is `less` (with `DVC_PAGER`, `PAGER` and `DEFAULT_PAGER` in order), and apply `-R` even if `$LESS` is specified (this should be applied without us doing anything and deviates from the behavior with `git`). And, if there's no `$LESS` set, we could use `less -FRSX`.

Otherwise we would need to disable the styles.

**Comment by jorgeorpinel** [comment]
I see #5906 addresses 1. from this issue but we still have 3-4 to decide on. Thanks!
